Implementation of ValidateAndGetBestMatch

alikindsys opened the pull request:
As discussed on #1699 this pr comes with a rewrite of the logic on `CommandService#ExecuteAsync`, splitting it into two functions.    
  
`ExecuteAsync` - will continue to do the `Search-Verify-Match-Execute` loop as it did before.  
  
`ValidateAndGetBestMatch` - Verifies from a set of commands (`SearchResult`) with the specified `ICommandContext`, `IServiceProvider` and `MultiMatchHandling` which command is the optimal one, using the previously established classification and ranking systems. 

Breaking Changes
---
None.

Focus
---
Code cleanup and creation of an new public api.

alikindsys committed:
Moves CalculateScore function to outer scope.

alikindsys committed:
Creates ValidateAndGetBestMatch function

This function will validate all commands from a SearchResult and return the result of said validation, along with the command matched, if a valid match was found.

alikindsys committed:
Make use of new ValidateAndGetBestMatch api on ExecuteAsync

alikindsys committed:
Fixes Azure linux build failing due to a CS8652.

quinchs commented:
Going to cc @Cenngo on this just for sanity

Cenngo commented on src/Discord.Net.Commands/CommandService.cs:
```diff
@@ -506,19 +506,66 @@ public async Task<IResult> ExecuteAsync(ICommandContext context, string input, I
+        /// <param name="multiMatchHandling">The handling mode when multiple command matches are found.</param>
+        /// <returns>A task that represents the asynchronous validation operation. The task result contains the result of the
+        ///     command validation and the command matched, if a match was found.</returns>
+        public async Task<(IResult, Optional<CommandMatch>)> ValidateAndGetBestMatch(SearchResult matches, ICommandContext context, IServiceProvider provider, MultiMatchHandling multiMatchHandling = MultiMatchHandling.Exception)
+        {
+            if (!matches.IsSuccess)
+                return (matches, Optional.Create<CommandMatch>());
```
I dont think returning an optional CommandMatch to indicate to the user that their SearchResult wasnt successful isnt the most user friendly way of validating a command search. Instead of returning a Tuple, i think you should encapsulate this in an `IResult` that way you can provide more information to the user and it would be easier to build a command execution pipeline around this, since your goal is to allow people to utilize this.

alikindsys replied:
I see.

Cenngo commented on src/Discord.Net.Commands/CommandService.cs:
```diff
@@ -506,19 +506,66 @@ public async Task<IResult> ExecuteAsync(ICommandContext context, string input, I
+            //Since ValidateAndGetBestMatch is deterministic on the return type, we can use pattern matching on the type for infering the code flow.
+            var (validationResult, commandMatch) = await ValidateAndGetBestMatch(searchResult, context, services, multiMatchHandling);
+
+            if(validationResult is SearchResult)
```
Tho i understand the way you are using this workflow to branch out the execution method, i dont think it is the most descript way of doing this. I think this should be more user readable to make it easier to maintain. Using an `IResult` to return the method result, with some additional info for branching out the pipeline, might be a better idea.

alikindsys replied:
Understood.

Cenngo commented:
Other than that, it all seems good.

quinchs commented:
@roridev Are you able to make the requested changes?

roridev commented:
Im able now, just give me a moment

alikindsys committed:
Add `MatchResult`

alikindsys committed:
Add requested changes

Changes:
- Use IResult instead of Optional CommandMatch

- Rework branching workflow

roridev commented:
Please check if the changes done align with what was requested.

Fun things that were discovered while doing this:

- A whole branch of execution was missing on my old code. The event wouldn't be fired in the case of  `ValidateAndGetBestMatch` returning a `SearchResult`. This is now fixed.

quinchs commented:
@Cenngo waiting on your final approval

Cenngo commented on src/Discord.Net.Commands/CommandService.cs:
```diff
@@ -506,19 +506,83 @@ public async Task<IResult> ExecuteAsync(ICommandContext context, string input, I
+            return validationResult;
+        }
+
+        private async Task<IResult> handleCommandPipeline(MatchResult matchResult, ICommandContext context, IServiceProvider services)
```
Could you please make this pascal casing?

alikindsys replied:
oh sure rider autocorrected it to that.

Cenngo reviewed (changes requested):
The changes you made address the issues i mentioned. Great to hear that this also led you to fixing a problem.

alikindsys committed:
Fix incorrect casing on `HandleCommandPipeline`
